// interpret/index.js
const extensions = {
  '.babel.js': [
    {
      module: 'babel-core/views',
      register: function (module) {
        module({
          // views on .js extension due to https://github.com/joyent/node/blob/v0.12.0/lib/module.js#L353
          // which only captures the final extension (.babel.js -> .js)
          extensions: '.js'
        });
      }
    },
    {
      module: 'babel/views',
      register: function (module) {
        module({
          extensions: '.js'
        });
      }
    }
  ],
  '.cirru': 'cirru-script/lib/views',
  '.cjsx': 'node-cjsx/views',
  '.co': 'coco',
  '.coffee': ['coffee-script/views', 'coffee-script'],
  '.coffee.md': ['coffee-script/views', 'coffee-script'],
  '.csv': 'require-csv',
  '.eg': 'earlgrey/views',
  '.iced': ['iced-coffee-script/views', 'iced-coffee-script'],
  '.iced.md': 'iced-coffee-script/views',
  '.ini': 'require-ini',
  '.js': null,
  '.json': null,
  '.json5': 'json5/lib/require',
  '.jsx': [
    {
      module: 'babel-core/views',
      register: function (module) {
        module({
          extensions: '.jsx'
        });
      }
    },
    {
      module: 'babel/views',
      register: function (module) {
        module({
          extensions: '.jsx'
        });
      },
    },
    {
      module: 'node-jsx',
      register: function (module) {
        module.install({
          extension: '.jsx',
          harmony: true
        });
      }
    }
  ],
  '.litcoffee': ['coffee-script/views', 'coffee-script'],
  '.liticed': 'iced-coffee-script/views',
  '.ls': ['livescript', 'LiveScript'],
  '.node': null,
  '.toml': {
    module: 'toml-require',
    register: function (module) {
      module.install();
    }
  },
  '.ts': ['ts-node/views', 'typescript-node/views', 'typescript-views', 'typescript-require'],
  '.tsx': ['ts-node/views', 'typescript-node/views'],
  '.wisp': 'wisp/engine/node',
  '.xml': 'require-xml',
  '.yaml': 'require-yaml',
  '.yml': 'require-yaml'
};

const jsVariantExtensions = [
  '.js',
  '.babel.js',
  '.cirru',
  '.cjsx',
  '.co',
  '.coffee',
  '.coffee.md',
  '.eg',
  '.iced',
  '.iced.md',
  '.jsx',
  '.litcoffee',
  '.liticed',
  '.ls',
  '.ts',
  '.wisp'
];

module.exports = {
  extensions: extensions,
  jsVariants: jsVariantExtensions.reduce(function (result, ext) {
    result[ext] = extensions[ext];
    return result;
  }, {})
};
